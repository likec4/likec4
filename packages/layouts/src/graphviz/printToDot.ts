/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { ComputedEdge, ComputedNode, ComputedView, Fqn } from '@likec4/core'
import { Colors, DefaultThemeColor, RelationColors, nameFromFqn, nonNullable } from '@likec4/core'
import { isTruthy } from 'remeda'
import {
  attribute as _,
  digraph,
  toDot,
  type $keywords,
  type GraphBaseModel,
  type NodeModel,
  type RootGraphModel,
  type SubgraphModel
} from 'ts-graphviz'
import { edgeLabel, nodeLabel, sanitize } from './dot-labels'
import { pxToInch, pxToPoints } from './graphviz-utils'
import type { DotSource } from './types'

// Declare custom attributes.
declare module 'ts-graphviz' {
  export namespace ClusterSubgraphAttributeKey {
    export interface $values extends $keywords<'likec4_id' | 'likec4_level' | 'likec4_depth'> {}
  }

  export namespace NodeAttributeKey {
    export interface $values extends $keywords<'likec4_id' | 'likec4_level'> {}
  }

  export namespace EdgeAttributeKey {
    export interface $values extends $keywords<'likec4_id'> {}
  }

  export namespace Attribute {
    export interface $keys extends $keywords<'likec4_id' | 'likec4_level' | 'likec4_depth'> {}

    export interface $types {
      likec4_id: string
      likec4_level: number
      likec4_depth: number
    }
  }
}

function isCompound(node: ComputedNode) {
  return node.children.length > 0
}

export function toGraphvisModel({ autoLayout, nodes, edges }: ComputedView): RootGraphModel {
  const G = digraph({
    [_.layout]: 'dot',
    [_.compound]: true,
    [_.pad]: pxToInch(10),
    [_.rankdir]: autoLayout,
    [_.outputorder]: 'nodesfirst',
    [_.nodesep]: pxToInch(80),
    [_.ranksep]: pxToInch(90),
    // [_.newrank]: true,
    [_.pack]: pxToPoints(20),
    [_.packmode]: 'array_t'
  })
  G.attributes.graph.apply({
    [_.fontname]: 'Helvetica',
    [_.labeljust]: 'l',
    [_.fontsize]: pxToPoints(12)
  })

  G.attributes.node.apply({
    [_.fontname]: 'Helvetica',
    [_.fontsize]: pxToPoints(19),
    [_.shape]: 'rect',
    [_.width]: pxToInch(320),
    [_.height]: pxToInch(180),
    [_.style]: 'filled,rounded',
    [_.fillcolor]: Colors[DefaultThemeColor].fill,
    [_.margin]: pxToInch(20),
    [_.penwidth]: 0
  })

  G.attributes.edge.apply({
    [_.fontname]: 'Helvetica',
    [_.fontsize]: pxToPoints(14),
    [_.style]: 'solid',
    [_.penwidth]: 2,
    [_.arrowsize]: 0.8,
    [_.color]: RelationColors.lineColor,
    [_.fontcolor]: RelationColors.labelColor
  })

  const subgraphs = new Map<Fqn, SubgraphModel>()
  const graphvizNodes = new Map<Fqn, NodeModel>()

  let sortv = 0
  const addNode = (elementNode: ComputedNode) => {
    let name = nameFromFqn(elementNode.id).toLowerCase()
    if (name.startsWith('cluster')) {
      name = 'nd_' + name
    }
    let id = name
    let uniqueNumber = 1
    while (G.getNode(id)) {
      id = name + '_' + uniqueNumber++
    }
    const node = G.createNode(id, {
      [_.likec4_id]: elementNode.id,
      [_.likec4_level]: elementNode.level,
      [_.label]: nodeLabel(elementNode),
      [_.sortv]: sortv++
    })
    if (elementNode.color !== DefaultThemeColor) {
      node.attributes.apply({
        [_.fillcolor]: Colors[elementNode.color].fill
      })
    }
    switch (elementNode.shape) {
      case 'queue': {
        node.attributes.apply({
          [_.width]: pxToInch(320),
          [_.height]: pxToInch(160)
        })
        break
      }
      case 'cylinder':
      case 'storage': {
        node.attributes.apply({
          [_.color]: Colors[elementNode.color].stroke,
          [_.penwidth]: 2,
          [_.shape]: 'cylinder'
        })
        break
      }
      default:
        break
    }
    graphvizNodes.set(elementNode.id, node)
    return node
  }

  /**
   * returns recursion depth
   */
  const traverseClusters = (
    elementNode: ComputedNode,
    parent: GraphBaseModel,
    level = 0
  ): number => {
    if (!isCompound(elementNode)) {
      const node = nonNullable(graphvizNodes.get(elementNode.id), "graphviz node doesn't exist")
      parent.node(node.id)
      return 0
    }
    const name = 'cluster_' + nameFromFqn(elementNode.id).toLowerCase()
    let id = name
    let uniqueNumber = 1
    while (G.getSubgraph(id)) {
      id = name + '_' + uniqueNumber++
    }

    const subgraph = parent.createSubgraph(id, {
      [_.likec4_id]: elementNode.id,
      [_.likec4_level]: level
    })

    subgraph.attributes.graph.apply({
      [_.fontsize]: pxToPoints(13),
      [_.style]: 'rounded',
      [_.margin]: pxToPoints(40)
    })

    const label = sanitize(elementNode.title.toUpperCase())
    if (isTruthy(label)) {
      subgraph.apply({
        [_.label]: `<<B>${label}</B>>`
      })
    }

    subgraphs.set(elementNode.id, subgraph)

    let depth = 0
    for (const child of nodes.filter(n => n.parent === elementNode.id)) {
      depth = Math.max(traverseClusters(child, subgraph, level + 1) + 1, depth)
    }
    subgraph.set(_.likec4_depth, depth)
    // TODO: calculate color here
    return depth
  }

  function addEdge<E extends ComputedEdge>(edge: E, parent: GraphBaseModel) {
    const source = graphvizNodes.get(edge.source)
    const target = graphvizNodes.get(edge.target)

    if (!source && !target) {
      return
    }
    let lhead, ltail: string | undefined
    // TODO: Edge with cluster?
    // if (!target) {
    //   const targetSubgraph = subgraphs.get(edge.target)
    //   if (!targetSubgraph?.id) {
    //     return
    //   }
    //   lhead = targetSubgraph.id
    //   const firstNode = head(
    //     nodes
    //       .filter(n => !isCompound(n) && isAncestor(edge.target, n.id))
    //       .sort((a, b) => b.inEdges.length - a.inEdges.length)
    //   )
    //   target = firstNode && graphvizNodes.get(firstNode.id)
    // }
    // if (!source) {
    //   const sourceSubgraph = subgraphs.get(edge.source)
    //   if (!sourceSubgraph?.id) {
    //     return
    //   }
    //   ltail = sourceSubgraph.id
    //   const firstNode = head(
    //     nodes
    //       .filter(n => !isCompound(n) && isAncestor(edge.source, n.id))
    //       .sort((a, b) => b.inEdges.length - a.inEdges.length)
    //   )
    //   source = firstNode && graphvizNodes.get(firstNode.id)
    // }

    if (source && target) {
      const e = parent.edge([source, target], {
        [_.likec4_id]: edge.id
        // [_.weight]: 20
      })

      const sourceLevel = source.attributes.get(_.likec4_level) ?? 0
      const targetLevel = target.attributes.get(_.likec4_level) ?? 0
      if (parent !== G) {
        e.attributes.apply({
          [_.weight]: sourceLevel === targetLevel ? 10 : 5
        })
      } else if (sourceLevel > 0 && sourceLevel == targetLevel) {
        // More weight for same level
        e.attributes.apply({
          [_.weight]: 5
        })
      }
      if (lhead) {
        e.attributes.apply({
          [_.lhead]: lhead
          // [_.weight]: 5
        })
      }
      if (ltail) {
        e.attributes.apply({
          [_.ltail]: ltail
          // [_.weight]: 5
        })
      }
      if (isTruthy(edge.label)) {
        const attr = lhead || ltail ? _.xlabel : _.label
        e.attributes.apply({
          [attr]: edgeLabel(edge.label),
          [_.nojustify]: true
        })
      }
    }
  }

  const clusters = [] as ComputedNode[]
  for (const node of nodes) {
    if (isCompound(node)) {
      if (!node.parent) {
        clusters.push(node)
      }
      continue
    }
    addNode(node)
  }
  let cluster
  while ((cluster = clusters.shift())) {
    traverseClusters(cluster, G)
  }

  for (const edge of edges) {
    const parent = (edge.parent && subgraphs.get(edge.parent)) ?? G
    addEdge(edge, parent)
  }

  return G
}

export function printToDot(view: ComputedView): DotSource {
  return toDot(toGraphvisModel(view)) as DotSource
}
