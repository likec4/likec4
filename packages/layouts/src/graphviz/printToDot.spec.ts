import { describe, expect, it } from 'vitest'
import {
  computedAmazonView,
  computedCloudView,
  computedCloud3levels,
  computedIndexView
} from './__fixtures__'
import { printToDot } from './printToDot'

describe('printToDot', () => {
  it('computedIndexView', () => {
    expect.hasAssertions()
    const diagram = printToDot(computedIndexView)
    expect(diagram).toMatchSnapshot()
  })

  it('computedAmazonView', () => {
    expect.hasAssertions()
    const diagram = printToDot(computedAmazonView)
    expect(diagram).toMatchSnapshot()
  })

  it('computedCloud3levels', () => {
    expect.hasAssertions()
    const diagram = printToDot(computedCloud3levels)
    expect(diagram).toMatchSnapshot()
  })

  it('computedCloudView', () => {
    expect.hasAssertions()
    const diagram = printToDot(computedCloudView)
    expect(diagram).toMatchSnapshot()
  })
})
