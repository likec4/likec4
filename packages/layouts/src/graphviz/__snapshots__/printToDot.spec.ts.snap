// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`printToDot > computedAmazonView 1`] = `
"digraph {
  layout = "dot";
  compound = true;
  pad = 0.1042;
  rankdir = "TB";
  outputorder = "nodesfirst";
  nodesep = 0.8334;
  ranksep = 0.9375;
  pack = 15;
  packmode = "array_t";
  graph [
    fontname = "Helvetica";
    labeljust = "l";
    fontsize = 9;
  ];
  edge [
    fontname = "Helvetica";
    fontsize = 10.5;
    style = "solid";
    penwidth = 2;
    arrowsize = 0.8;
    color = "#b1b1b1";
    fontcolor = "#b1b1b1";
  ];
  node [
    fontname = "Helvetica";
    fontsize = 14.25;
    shape = "rect";
    width = 3.3334;
    height = 1.875;
    style = "filled,rounded";
    fillcolor = "#3b82f6";
    margin = 0.2084;
    penwidth = 0;
  ];
  "backend" [
    likec4_id = "cloud.backend";
    likec4_level = 1;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">Backend</FONT>>;
    sortv = 0;
  ];
  "s3" [
    likec4_id = "amazon.s3";
    likec4_level = 1;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">S3</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">S3 is a storage service</FONT></TD></TR></TABLE>>;
    sortv = 1;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    label = <<B>CLOUD</B>>;
    likec4_depth = 1;
    graph [
      fontsize = 9.75;
      style = "rounded";
      margin = 30;
    ];
    "backend";
  }
  subgraph "cluster_amazon" {
    likec4_id = "amazon";
    likec4_level = 0;
    label = <<B>AMAZON</B>>;
    likec4_depth = 1;
    graph [
      fontsize = 9.75;
      style = "rounded";
      margin = 30;
    ];
    "s3";
  }
  "backend" -> "s3" [
    likec4_id = "cloud.backend:amazon.s3";
    weight = 5;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">persists artifacts</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
}"
`;

exports[`printToDot > computedCloud3levels 1`] = `
"digraph {
  layout = "dot";
  compound = true;
  pad = 0.1042;
  rankdir = "TB";
  outputorder = "nodesfirst";
  nodesep = 0.8334;
  ranksep = 0.9375;
  pack = 15;
  packmode = "array_t";
  graph [
    fontname = "Helvetica";
    labeljust = "l";
    fontsize = 9;
  ];
  edge [
    fontname = "Helvetica";
    fontsize = 10.5;
    style = "solid";
    penwidth = 2;
    arrowsize = 0.8;
    color = "#b1b1b1";
    fontcolor = "#b1b1b1";
  ];
  node [
    fontname = "Helvetica";
    fontsize = 14.25;
    shape = "rect";
    width = 3.3334;
    height = 1.875;
    style = "filled,rounded";
    fillcolor = "#3b82f6";
    margin = 0.2084;
    penwidth = 0;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">customer</FONT>>;
    sortv = 0;
  ];
  "dashboard" [
    likec4_id = "cloud.frontend.dashboard";
    likec4_level = 1;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Customer Dashboard Webapp</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">The customer dashboard is a webapp that<BR/>allows customers to view their data.<BR/></FONT></TD></TR></TABLE>>;
    sortv = 1;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Support Engineer</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Support engineers are responsible for<BR/>supporting customers<BR/></FONT></TD></TR></TABLE>>;
    sortv = 2;
  ];
  "adminpanel" [
    likec4_id = "cloud.frontend.adminPanel";
    likec4_level = 1;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Admin Panel Webapp</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">The admin panel is a webapp that allows<BR/>support staff to manage customer data.<BR/></FONT></TD></TR></TABLE>>;
    sortv = 3;
  ];
  "graphql" [
    likec4_id = "cloud.backend.graphql";
    likec4_level = 2;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Graphql API</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Component that allows to query data via<BR/>GraphQL.<BR/></FONT></TD></TR></TABLE>>;
    sortv = 4;
  ];
  "storage" [
    likec4_id = "cloud.backend.storage";
    likec4_level = 2;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Backend Storage</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">The backend storage is a component that<BR/>stores data.<BR/></FONT></TD></TR></TABLE>>;
    sortv = 5;
    color = "#2563eb";
    penwidth = 2;
    shape = "cylinder";
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Amazon</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
    sortv = 6;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    label = <<B>CLOUD</B>>;
    likec4_depth = 2;
    graph [
      fontsize = 9.75;
      style = "rounded";
      margin = 30;
    ];
    "dashboard";
    "adminpanel";
    subgraph "cluster_backend" {
      likec4_id = "cloud.backend";
      likec4_level = 1;
      label = <<B>BACKEND</B>>;
      likec4_depth = 1;
      graph [
        fontsize = 9.75;
        style = "rounded";
        margin = 30;
      ];
      "graphql";
      "storage";
      "graphql" -> "storage" [
        likec4_id = "cloud.backend.graphql:cloud.backend.storage";
        weight = 10;
      ];
    }
    "adminpanel" -> "graphql" [
      likec4_id = "cloud.frontend.adminPanel:cloud.backend.graphql";
      weight = 5;
      label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">fetches data in zero trust network with sso<BR/>authentification<BR ALIGN="LEFT"/></FONT></TD></TR></TABLE>>;
      nojustify = true;
    ];
    "dashboard" -> "graphql" [
      likec4_id = "cloud.frontend.dashboard:cloud.backend.graphql";
      weight = 5;
      label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">fetches data</FONT></TD></TR></TABLE>>;
      nojustify = true;
    ];
  }
  "storage" -> "amazon" [
    likec4_id = "cloud.backend.storage:amazon";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">persists artifacts</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "support" -> "adminpanel" [
    likec4_id = "support:cloud.frontend.adminPanel";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">manages</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "customer" -> "dashboard" [
    likec4_id = "customer:cloud.frontend.dashboard";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">opens in browser</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
}"
`;

exports[`printToDot > computedCloudView 1`] = `
"digraph {
  layout = "dot";
  compound = true;
  pad = 0.1042;
  rankdir = "TB";
  outputorder = "nodesfirst";
  nodesep = 0.8334;
  ranksep = 0.9375;
  pack = 15;
  packmode = "array_t";
  graph [
    fontname = "Helvetica";
    labeljust = "l";
    fontsize = 9;
  ];
  edge [
    fontname = "Helvetica";
    fontsize = 10.5;
    style = "solid";
    penwidth = 2;
    arrowsize = 0.8;
    color = "#b1b1b1";
    fontcolor = "#b1b1b1";
  ];
  node [
    fontname = "Helvetica";
    fontsize = 14.25;
    shape = "rect";
    width = 3.3334;
    height = 1.875;
    style = "filled,rounded";
    fillcolor = "#3b82f6";
    margin = 0.2084;
    penwidth = 0;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">customer</FONT>>;
    sortv = 0;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Support Engineer</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Support engineers are responsible for<BR/>supporting customers<BR/></FONT></TD></TR></TABLE>>;
    sortv = 1;
  ];
  "frontend" [
    likec4_id = "cloud.frontend";
    likec4_level = 1;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">Frontend</FONT>>;
    sortv = 2;
  ];
  "backend" [
    likec4_id = "cloud.backend";
    likec4_level = 1;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">Backend</FONT>>;
    sortv = 3;
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Amazon</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
    sortv = 4;
  ];
  subgraph "cluster_cloud" {
    likec4_id = "cloud";
    likec4_level = 0;
    label = <<B>CLOUD</B>>;
    likec4_depth = 1;
    graph [
      fontsize = 9.75;
      style = "rounded";
      margin = 30;
    ];
    "frontend";
    "backend";
    "frontend" -> "backend" [
      likec4_id = "cloud.frontend:cloud.backend";
      weight = 10;
      label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">fetches data</FONT></TD></TR></TABLE>>;
      nojustify = true;
    ];
  }
  "backend" -> "amazon" [
    likec4_id = "cloud.backend:amazon";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">persists artifacts</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "support" -> "frontend" [
    likec4_id = "support:cloud.frontend";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">manages</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "customer" -> "frontend" [
    likec4_id = "customer:cloud.frontend";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">opens in browser</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
}"
`;

exports[`printToDot > computedIndexView 1`] = `
"digraph {
  layout = "dot";
  compound = true;
  pad = 0.1042;
  rankdir = "TB";
  outputorder = "nodesfirst";
  nodesep = 0.8334;
  ranksep = 0.9375;
  pack = 15;
  packmode = "array_t";
  graph [
    fontname = "Helvetica";
    labeljust = "l";
    fontsize = 9;
  ];
  edge [
    fontname = "Helvetica";
    fontsize = 10.5;
    style = "solid";
    penwidth = 2;
    arrowsize = 0.8;
    color = "#b1b1b1";
    fontcolor = "#b1b1b1";
  ];
  node [
    fontname = "Helvetica";
    fontsize = 14.25;
    shape = "rect";
    width = 3.3334;
    height = 1.875;
    style = "filled,rounded";
    fillcolor = "#3b82f6";
    margin = 0.2084;
    penwidth = 0;
  ];
  "customer" [
    likec4_id = "customer";
    likec4_level = 0;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">customer</FONT>>;
    sortv = 0;
  ];
  "support" [
    likec4_id = "support";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Support Engineer</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Support engineers are responsible for<BR/>supporting customers<BR/></FONT></TD></TR></TABLE>>;
    sortv = 1;
  ];
  "cloud" [
    likec4_id = "cloud";
    likec4_level = 0;
    label = <<FONT COLOR="#eff6ff" POINT-SIZE="14.25">cloud</FONT>>;
    sortv = 2;
  ];
  "amazon" [
    likec4_id = "amazon";
    likec4_level = 0;
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="0" CELLSPACING="7"><TR><TD><FONT COLOR="#eff6ff" POINT-SIZE="14.25">Amazon</FONT></TD></TR><TR><TD><FONT COLOR="#bfdbfe" POINT-SIZE="10.5">Amazon is a cloud provider</FONT></TD></TR></TABLE>>;
    sortv = 3;
  ];
  "cloud" -> "amazon" [
    likec4_id = "cloud:amazon";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">persists artifacts</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "support" -> "cloud" [
    likec4_id = "support:cloud";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">manages</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
  "customer" -> "cloud" [
    likec4_id = "customer:cloud";
    label = <<TABLE BORDER="0" CELLBORDER="0" CELLPADDING="3" CELLSPACING="0"><TR><TD><FONT POINT-SIZE="10.5">opens in browser</FONT></TD></TR></TABLE>>;
    nojustify = true;
  ];
}"
`;
